### ISSUE TITLE
busted isn't running all tests?

### ISSUE THREAD

**Issue by daurnimator** [Olivine-Labs/busted#446]
Running `busted` for one of my projects doesn't seem to be running all tests...
```
$ busted 
●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●
41 successes / 0 failures / 0 errors / 0 pending : 0.133907 seconds
$ busted -l | wc -l
62
$ busted -c
●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●●
62 successes / 0 failures / 0 errors / 0 pending : 0.84043 seconds
```

Plain `busted` is only running 41 of my tests; whereas `busted -c` is running all 62. `busted -l` manages to see all 62.

**Comment by daurnimator** [comment]
This may be caused by an error thrown from a `__gc` metamethod.

**Comment by daurnimator** [comment]
> This may be caused by an error thrown from a __gc metamethod.

Was reading through your code, https://github.com/Olivine-Labs/busted/blob/master/busted/core.lua#L176 is suspicious:
an error in a `__gc` metamethod won't call an `xpcall` handler. Which means that `ret[1]` will be false, but `status` will be `"success"`.

```
> xpcall(function() setmetatable({}, {__gc=function() error("fooasd") end}); collectgarbage() collectgarbage() end, print)
false	error in __gc metamethod (stdin:1: fooasd)
```

**Comment by o-lim** [comment]
I submitted a PR to fix the gc issue, but I'm not sure it will help with the problem you are seeing.  Do you think you can narrow down what tests are not being run?  Is there a block or set of files that busted is not running?

**Comment by DorianGray** [comment]
will close as soon as new rc is released and fix confirmed

**Comment by daurnimator** [comment]
In the mean-time I've tracked down + fixed the original bug that was causing this... and now I don't have a replicator :(

**Comment by o-lim** [comment]
Could you describe the original bug that was causing this?

**Comment by daurnimator** [comment]
> Could you describe the original bug that was causing this?

When writing code I sometimes add `__gc` metamethods that throw an error if the object is not in a safe state to be discarded; handy for catching programming errors when prototyping.

I had messed up the internal state of an object; and it was (rightfully) throwing when it finally got collected. However, busted didn't cope well with this, and only when I **really** paid attention did I notice that some of my tests were mysteriously missing.

Removing the `__gc` metamethod made everything pass; so I attributed the issue to busted not handling `LUA_ERRGCMM` well, and went reading :P

In the mean time, I was working on my (untracked/uncommited) prototype code, and the original bug which was throwing the error in `__gc` got fixed.
I tried to recreate the circumstances (with old busted release), but I think it very much depends on when the garbage collector runs; and I haven't been able to get it in the correct spot again.
